Handle large IPv6 headers properly

rveerama1 opened the pull request:
Current implementation does not handle large extension headers (e.g HBHO). Which resulted network stack crashes or due to misinterpretation of lengths network packets are dropped. Parsing errors fixed. Also fixed sending side issues. (e.g If stack receives large HBHO with large echo packet, IPv6 fragmentation failed to build such replies).  Added more unit tests.

jukkar commented on include/net/net_pkt.h:
```diff
@@ -128,7 +128,7 @@ struct net_pkt {
 	};
```
Typo in commit message of "net: ipv6: Fix HBHO parsing options function"

s/parsing is failed/parsing fails/
s/declared as/is declared as/

The "PAD1 contains only one byte which is 0." looks confusing, could we say here "PAD1 contains only one byte which has value 0."

jukkar commented on subsys/net/ip/ipv6.c:
```diff
@@ -4251,15 +4261,14 @@ enum net_verdict net_ipv6_process_pkt(struct net_pkt *pkt)
 			NET_DBG("IPv6 next header %d", next);
-			net_pkt_set_ipv6_ext_len(pkt, offset -
-						 sizeof(struct net_ipv6_hdr));
+			net_pkt_set_ipv6_ext_len(pkt, length);
```
As there can be multiple extension headers, we need to accumulate the total length here.

jukkar commented on samples/net/echo_server/src/common.h:
```diff
@@ -16,7 +16,8 @@ void stop_tcp(void);
 
```
Typos in commit message of "samples: net: echo_server: Fix handling of large IPv6 header"

s/single buffer net_buf_pull/single buffer then net_buf_pull/
s/and pull of the header/and pull the header/

The last sentence needs clarification, perhaps something like "After this, the packet payload can be copied correctly when replying messages."

jukkar commented on samples/net/echo_server/src/echo-server.c:
```diff
@@ -49,12 +49,44 @@ void quit(void)
+			temp->frags = NULL;
+
+			net_buf_unref(temp);
+
```
Remove extra empty line

jukkar commented on samples/net/echo_server/src/echo-server.c:
```diff
@@ -70,21 +102,17 @@ struct net_pkt *build_reply_pkt(const char *name,
-		 * data.
-		 */
-		net_buf_pull(tmp, header_len);
+		NET_DBG("pull of header %d", header_len);
```
This will print too much information as debugging is enabled by default. Thus this print can be removed.

jukkar commented on samples/net/echo_server/src/echo-server.c:
```diff
@@ -70,21 +102,17 @@ struct net_pkt *build_reply_pkt(const char *name,
-		 */
-		net_buf_pull(tmp, header_len);
+		NET_DBG("pull of header %d", header_len);
+		pull_of_header(pkt, header_len);
```
The "of" looks redundant information, what about just calling this "pull_header()" instead.

rveerama1 commented:
I have noticed other issues also. I am working on it.

rveerama1 committed:
net: ipv6: Handle large IPv6 packets properly

Current implementation does not handle large extension headers
(e.g HBHO). Which resulted network stack crashes or due to
misinterpretation of lengths network packets are dropped. Also
caused issues while preparing IPv6 packet (e.g. large HBHO header
with IPv6 fragmentation support).

Issues fixed and provided more unit tests.

Signed-off-by: Ravi kumar Veeramally <ravikumar.veeramally@linux.intel.com>

rveerama1 committed:
samples: net: echo_server: Fix building of echo replies

If echo-server receives a IPv6 packet with large header (e.g HBHO)
build_reply_pkt() function failed to remove IPv6 header. Reason is
net_buf_pull() doesn't work if header length is more than one
fragment. net_pkt_pull() solves the issue.

Signed-off-by: Ravi kumar Veeramally <ravikumar.veeramally@linux.intel.com>

jukkar reviewed (approved):
LGTM
